File: programmers/_72415_flip_card.java
import java.util.*;

public class _72415_flip_card {

    static int[] dy = {-1, 1, 0, 0};
    static int[] dx = {0, 0, -1, 1};

    static class Point {

    }

    static class CardNode {

        int y, x;
        boolean isFlip;
        int f_y, f_x;

        @Override
        public String toString() {
            return "CardNode{" +
                "y=" + y +
                ", x=" + x +
                ", isFlip=" + isFlip +
                ", f_y=" + f_y +
                ", f_x=" + f_x +
                ", board='" + board + '\'' +
                '}';
        }

        String board;

        public CardNode(int y, int x, boolean isFlip, int f_y, int f_x, String board) {
            this.y = y;
            this.x = x;
            this.isFlip = isFlip;
            this.f_y = f_y;
            this.f_x = f_x;
            this.board = board;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }

            CardNode cardNode = (CardNode) o;

            if (y != cardNode.y) {
                return false;
            }
            if (x != cardNode.x) {
                return false;
            }
            if (isFlip != cardNode.isFlip) {
                return false;
            }
            if (f_y != cardNode.f_y) {
                return false;
            }
            if (f_x != cardNode.f_x) {
                return false;
            }
            return board.equals(((CardNode) o).board);
        }

        @Override
        public int hashCode() {
            int result = y;
            result = 31 * result + x;
            result = 31 * result + (isFlip ? 1 : 0);
            result = 31 * result + f_y;
            result = 31 * result + f_x;
            result = 31 * result + (board != null ? board.hashCode() : 0);
            return result;
        }
    }

    static class CardNodeWithCount {

        CardNode cardNode;
        int cnt;

        public CardNodeWithCount(CardNode cardNode, int cnt) {
            this.cardNode = cardNode;
            this.cnt = cnt;
        }
    }

    static int convertIndex(int y, int x) {
        return 4 * y + x;
    }

    static String switchToZeroFromChar(String board, char num) {
        StringBuilder temp = new StringBuilder(board);
        return temp.toString().replace(num, '0');
    }

    static String serialize(int[][] board) {
        StringBuilder builder = new StringBuilder();
        for (int[] row : board) {
            for (int num : row) {
                builder.append(num);
            }
        }

        return builder.toString();
    }

    static void printBoard(int[][] board) {
        for (int[] row : board) {
            StringBuilder builder = new StringBuilder();
            for (int num : row) {
                builder.append(num);
            }
            System.out.println(builder.toString());
        }
    }

    static boolean isEnd(String board) {
        for (int i = 0; i < board.length(); i++) {
            if (board.charAt(i) != '0') {
                return false;
            }
        }
        return true;
//        return board.indexOf('0') != -1;
    }

    static boolean isOOB(int y, int x) {
        return y >= 4 || y < 0 || x >= 4 || x < 0;
    }

    static boolean isOnEdgeByDirection(int y, int x, int d) {
        switch (d) {
            case 0:
                if (y == 0) {
                    return true;
                }
                break;
            case 1:
                if (y == 3) {
                    return true;
                }
                break;
            case 2:
                if (x == 0) {
                    return true;
                }
                break;
            case 3:
                if (x == 3) {
                    return true;
                }
                break;
            default:
                return false;
        }
        return false;
    }


    //    static boolean[][][] visit;
    static Set<CardNode> visit;


    public static int solution(int[][] board, int r, int c) {
        int answer = 40;

//        visit = new boolean[4][4][2];

        visit = new HashSet<>();

        ArrayDeque<CardNodeWithCount> q = new ArrayDeque<>();

        CardNode e = new CardNode(r, c, false, -1, -1, serialize(board));
        CardNodeWithCount e1 = new CardNodeWithCount(e, 0);
        q.addLast(e1);

        visit.add(e);

        while (!q.isEmpty()) {
            CardNodeWithCount cardNodeWithCount = q.pollFirst();
            CardNode poll = cardNodeWithCount.cardNode;
//            System.out.println(poll.board);


            for (int k = 0; k < 4; k++) {
                int ny = poll.y + dy[k];
                int nx = poll.x + dx[k];

                if (isOOB(ny, nx)) {
                    continue;
                }

                // ctrl 이동 , 0이 아닌곳을 만나면 멈춰야함
                while (poll.board.charAt(convertIndex(ny, nx)) == '0' && !isOnEdgeByDirection(ny,
                    nx, k)) {
                    ny = ny + dy[k];
                    nx = nx + dx[k];
                }

                CardNode newCardPair = new CardNode(ny, nx, poll.isFlip, poll.f_y,
                    poll.f_x, poll.board);

                if (visit.contains(newCardPair)) {
                    continue;
                }

                visit.add(newCardPair);
//                q.add();
                q.addLast(new CardNodeWithCount(newCardPair, cardNodeWithCount.cnt + 1));
            }

            for (int k = 0; k < 4; k++) {
                int ny = poll.y + dy[k];
                int nx = poll.x + dx[k];

                if (isOOB(ny, nx)) {
                    continue;
                }

                CardNode newCardPair = new CardNode(ny, nx, poll.isFlip, poll.f_y,
                    poll.f_x, poll.board);

                if (visit.contains(newCardPair)) {
                    continue;
                }

                visit.add(newCardPair);
                q.addLast(new CardNodeWithCount(newCardPair, cardNodeWithCount.cnt + 1));
            }

            if (poll.isFlip) {
                if (poll.board.charAt(convertIndex(poll.f_y, poll.f_x)) == poll.board.charAt(
                    convertIndex(poll.y, poll.x)) && !(poll.f_y==poll.y&&poll.f_x==poll.x)) {

                    String newBoard = switchToZeroFromChar(poll.board,
                        poll.board.charAt(convertIndex(poll.f_y, poll.f_x)));

                    if (isEnd(newBoard)) {
                        return cardNodeWithCount.cnt + 1;
                    }

                    CardNode newCardPair = new CardNode(poll.y, poll.x, false, -1, -1,
                        newBoard);

                    visit.add(newCardPair);
                    q.addLast(new CardNodeWithCount(newCardPair, cardNodeWithCount.cnt + 1));

                } else {
                    CardNode newCardPair = new CardNode(poll.y, poll.x, false, -1, -1,
                        poll.board);

                    if (visit.contains(newCardPair)) {
                        continue;
                    }

                    visit.add(newCardPair);
                    q.addLast(new CardNodeWithCount(newCardPair, cardNodeWithCount.cnt + 1));

                }

            } else {
                CardNode newCardPair = new CardNode(poll.y, poll.x, true, poll.y,
                    poll.x, poll.board);

                if (visit.contains(newCardPair)) {
                    continue;
                }

                visit.add(newCardPair);
                q.addLast(new CardNodeWithCount(newCardPair, cardNodeWithCount.cnt + 1));


            }
        }

        return answer;


    }

    public static void main(String[] args) {
        int[][] board = {{0, 0, 1, 0}, {1, 0, 0, 0}, {4, 4, 3, 2}, {0, 3, 2, 0}};
        System.out.println(solution(board, 2, 0));
    }

}

File: programmers/72415_flip_card.py
from collections import deque

dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]


def solution(board, r, c):
    def isEnd(s) -> bool:
        for c in s:
            if c != "0":
                return False
        return True

    def OOB(y, x) -> bool:
        if (not 0 <= y < 4) or (not 0 <= x < 4):
            return True
        return False

    def isOnEdgeByDirection(y, x, d) -> bool:

        if d == 0:
            if y == 0:
                return True
        elif d == 1:
            if y == 3:
                return True
        elif d == 2:
            if x == 0:
                return True
        elif d == 3:
            if x == 3:
                return True
        return False

    # 바뀌는 상황 :
    # 카드판의 상황,현재 커서의 위치 , 현재 뒤집었는지? 아니면 안뒤집엇는지 ( 2장째 기다리는중인지 뒤집었다면 커서 위치)

    # 구하는 것 : 최소 회수 이동

    def serialize(board) -> str:
        ret = ""
        for r in board:
            for num in r:
                ret += str(num)

        return ret

    def switchTo0(s: str, num: str) -> int:

        return s.replace(num, "0")

    def idxConverter(y, x) -> int:
        return 4 * y + x

    visit = set()

    # 보드의 상태 ,커서의 위치 , 커서가 카드를 뒤집었는지 , 뒤집힌 카드가 있다면 그 위치
    visit.add((serialize(board), (r, c), False, (-1, -1)))
    q = deque()
    q.append((serialize(board), r, c, False, -1, -1, 0))

    while q:
        b, y, x, isFlipped, f_y, f_x, count = q.popleft()

        # 모든 카드를 뒤집었는지
        # 여기서 체크하면 시간 더걸림
        # if isEnd(b):
        #     return count

        for k in range(4):
            ny, nx = y + dy[k], x + dx[k]

            if OOB(ny, nx):
                continue

            while b[idxConverter(ny, nx)] == "0" and not isOnEdgeByDirection(ny, nx, k):
                ny, nx = ny + dy[k], nx + dx[k]

            if (b, (ny, nx), isFlipped, (f_y, f_x)) in visit:
                continue

            visit.add((b, (ny, nx), isFlipped, (f_y, f_x)))
            q.append((b, ny, nx, isFlipped, f_y, f_x, count + 1))

        # 상하좌우 네방향 이동

        for k in range(4):
            ny, nx = y + dy[k], x + dx[k]

            if OOB(ny, nx):
                continue

            if (b, (ny, nx), isFlipped, (f_y, f_x)) in visit:
                continue

            visit.add((b, (ny, nx), isFlipped, (f_y, f_x)))
            q.append((b, ny, nx, isFlipped, f_y, f_x, count + 1))

        # 현재 한장은 뒤집은 상태에서 다른것 뒤집으려 할때
        if isFlipped:
            # 카드 짝 찾는데 성공하면
            # 0으로 뒤집고
            if b[idxConverter(f_y, f_x)] == b[idxConverter(y, x)] and (f_y, f_x) != (y, x):

                b = switchTo0(b, b[idxConverter(f_y, f_x)])

                # 여기서 끝낼수있음

                if isEnd(b):
                    return count + 1

                # print(b)
                visit.add((b, (y, x), False, (-1, -1)))
                q.append((b, y, x, False, -1, -1, count + 1))

            # 짝은 안맞으면
            else:

                if (b, (y, x), False, (-1, -1)) in visit:
                    continue
                visit.add((b, (y, x), False, (-1, -1)))
                q.append((b, y, x, False, -1, -1, count + 1))
        # 아직 뒤집지 않은 상태
        # 이제 뒤집기
        else:
            if (b, (y, x), True, (y, x)) in visit:
                continue

            visit.add((b, (y, x), True, (y, x)))
            q.append((b, y, x, True, y, x, count + 1))


# print(solution([[1, 0, 0, 3], [2, 0, 0, 0], [0, 0, 0, 2], [3, 0, 1, 0]], 1, 0))

# print(solution([[1, 5, 0, 2], [6, 4, 3, 0], [0, 2, 1, 5], [3, 0, 6, 4]], 0, 0))
